aws/rust-runtime/aws-inlineable/src/http_request_checksum.rs

```diff
@@ -60,6 +60,45 @@ impl Storable for RequestChecksumInterceptorState {
     type Storer = StoreReplace<Self>;
 }
 
+pub(crate) type CustomDefaultFn = Box<
+    dyn Fn(Option<ChecksumAlgorithm>, &ConfigBag) -> Option<ChecksumAlgorithm>
+        + Send
+        + Sync
+        + 'static,
+>;
+
+pub(crate) struct DefaultRequestChecksumOverride {
```

Comment:
I partially implemented that approach but abandoned it because I suspected that it was more complicated. The challenge is that if we have an S3-specific `RequestChecksumInterceptor`, then it needs to either
- disable [the general RequestChecksumInterceptor](https://github.com/smithy-lang/smithy-rs/blob/f7be5f801504e7edf5cb6b731f3400c37b298abc/aws/rust-runtime/aws-inlineable/src/http_request_checksum.rs#L79-L100) entirely
- only change the checksum default and delegate the rest to the general `RequestChecksumInterceptor`.

For the former, all the logic in the general `RequestChecksumInterceptor` needs to be taken over by the S3-specific `RequestChecksumInterceptor`, but the general one must be disabled in runtime components, otherwise a request checksum related header will incorrectly be added twice (plus there is a problem of code duplication between the general `RequestChecksumInterceptor` and S3-specifc one).

For the latter, we need to have a decorating interceptor so the S3-specific `RequestChecksumInterceptor` wraps the current `RequestChecksumInterceptor`. But then again, we must disable the general `RequestChecksumInterceptor` for the same reason mentioned above.

To disable the general `RequestChecksumInterceptor` using [disable_interceptor](https://github.com/smithy-lang/smithy-rs/blob/main/rust-runtime/aws-smithy-runtime-api/src/client/interceptors.rs#L839), we need to remove a generic parameter `AP` from `RequestChecksumInterceptor` because we cannot enumerate all possible `AP` when specifying it in `disable_interceptor`.

Let me know if I'm overthinking and you have something simpler.

Reply:
I see. I neglected to consider the complex interplay between the Smithy/AWS codegen pieces. That makes sense.